File: Problems/1518. (E) Water Bottles/Python3/solution.py
class Solution:
  def numWaterBottles(self, numBottles: int, numExchange: int) -> int:
    # simulation
    
    ret = 0
    empty, full = 0, numBottles
    while full + empty >= numExchange:
      # drink full bottles
      ret += full
      # recompute number of each bottle
      full, empty = 0, full + empty
      full, empty = empty // numExchange, empty % numExchange

    return ret + full

File: Problems/1518. (E) Water Bottles/JavaScript/solution.js
var numWaterBottles = function(numBottles, numExchange) {
  // simulation
  let res = 0;
  let empty = 0;
  while (numBottles > 0) {
    // drink bottle and update values accordingly
    res++;
    empty++;
    numBottles--;
    // exchange empty bottles when possible
    if (empty === numExchange) {
      empty = 0;
      numBottles++;
    }
  }
  return res;
};

